### ISSUE TITLE
Incremental Changelog generation

### ISSUE THREAD

**Issue by rdbisme** [python-semantic-release/python-semantic-release#722]
Hello, I'm trying to integrate this tool to our big repo (~3k commits) and the changelog generation takes a lot of time (40m). Is there a way to generate the changelog incrementally (once generated for all the commits) that just appends the missing part of it without rewalking the entire git tree?

**Comment by bernardcooke53** [comment]
Hey @rdbisme - there isn't currently, that gets really tricky when dealing with multiple branches where some changes are/aren't on the current branch. I think it could be possible to do with a comment in the changelog indicating the "last" version that it was generated for, but could still be broken pretty easily with a custom template.

I think it's potentially doable but a non-trivial amount of work

**Comment by Sieboldianus** [comment]
I agree. I had a careflly crafted CHANGELOG that got overwritten since v8 was release. I wish there was such a `last` option to indicate a commit below which Semantic-release should not delete entries. This could also be used to once-in-a-while walk through the changelog and manually clean things up.

**Comment by codejedi365** [comment]
No update here yet, however, I have been evaluating performance improvements across the project and this example is one that I use to theorize improvements. 

My latest idea is to add caching of the parsed commit objects which in theory should never need to be parsed again unless the configuration significantly changes or you rebase the entire project. This should decrease all subsequent runtimes with caching set up in the CIs.

I will keep evaluating to see if a partial git tree evaluation can be added but as it was stated above, definitely a lot of work to accomplish.

**Comment by codejedi365** [comment]
Unfortunately, no movement here yet

**Comment by codejedi365** [comment]
No significant movement here yet with performance improvements. It is constantly a use case that I consider regularly, so thank you for providing an example.

**Comment by codejedi365** [comment]
@rdbisme, I wanted to let you know that with `v9.10.0` the changelog can now be updated via the "update" mode. Check out the Changelog Templates in the documentation to learn about the "update" mode.

This change does not remove the need to re-walk the git tree but it might give you a faster runtime. Either way I plan to keep this issue open as I work to increase speed of PSR.

**Comment by rdbisme** [comment]
Thank you. I'll give it a look. 

Actually I've grown accustomed to this being stateless. It has its advantages. Maybe you can parallelize the commit walk? Should speed up stuff quite considerably.

**Comment by codejedi365** [comment]
@rdbisme, I'm not sure what benefits you are referring to when related to being stateless for this situation.

Separately, have you tried just turning down the verbosity. PSR's repo is about half the size of yours now and I see a ridiculous change in speed when I'm at `-vv` versus none. I think it's because of the Python logging module that is logging during the commit walk. If you find this is the case, then I'll look into refining the log statements more so recommended/normal evaluation is at a lower verbosity level

**Comment by codejedi365** [comment]
Still in the backlog

**Comment by codejedi365** [comment]
Still in backlog

**Comment by codejedi365** [comment]
Currently more significant improvement in work likely for the next major release. I would be curious if you have already seen an improvement in speed lately?

Also with your answer, please let me know if you are using the GitHub Docker Action or are you using it directly. You will find that if you use python-semantic-release directly (as opposed to the Docker action) you will get significant GitHub Actions workflow execution time decreases.

**Comment by codejedi365** [comment]
Still in backlog
